### PR TITLE
Added support for Quark I2C interfaces: I2C0 and I2C1.

### PR DESCRIPTION
**Pull request by SidLeung**
Additions:
1.  Ported Quark I2C driver to ARC/CoreLibs, located at system/libarc32_arduino101/drivers.
2.  Created the I2C Adapters, located at cores/arduino.

Modifications:
1.  Platform.h - added number of I2C interface definition.

### CODE REVIEW

**Comment by eriknyquist** [comment]
@SidLeung is it correct to say that all the files in cores/arduino are brand new files?

**Comment by SidLeung** [comment]
The additional files are ported from TD, I believe.

**Comment by eriknyquist** [comment]
Is this related to PR #218?

**Comment by SidLeung** [comment]
Yes.  PR #218 is going to be closed without merging back to the main trunk.  Actually, thank you for reminding me to do that...

**Comment by SidLeung** [comment]
@bbaltz505 Would you please close out PR-218 without merging it.  We are going on this branch.

**Comment by SidLeung** [comment]
@bigdinotech @eriknyquist  Russ has been testing this branch of Quark I2C driver together with library and sketch in the Extended Support Library.  Please review the code.  Thank you.  @russmcinnis Please give a brief update on the testing of this code so far.  Thanks.

**Comment by bigdinotech** [comment]
@SidLeung I don't think the example test sketch belongs here since it is just for testing and may just confuse users. Otherwise. +1

File: cores/arduino/soc_i2c.c [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+    return ret;
+}
+
+void soc_i2c_close_adapter(SOC_I2C_CONTROLLER controller_id)
+{
+    soc_i2c_deconfig(controller_id);
+    soc_i2c_clock_disable(controller_id);
+
+    SET_PIN_MODE(20, GPIO_MUX_MODE);
```
Comment (xieqi) [inline cores/arduino/soc_i2c.c]: it should according to the I2C controller id to set different pins, for I2C0, the pins are 20 and 21, for I2C1 the pins are 22 and 23

Comment (SidLeung) [inline cores/arduino/soc_i2c.c]: That's a bug.  Will fix it.  BTW should the pin be set to high, Qi?  @xieqi

File: libraries/Wire/examples/master_check/master_check.ino [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+static int count = 0;  // recode the higher 4 bits of data
+static uint8_t buffer_sender[BUFFER_SIZE]; // data source for master writer
+static uint8_t buffer_receiver[BUFFER_SIZE]; // data distination for master reader
+
+void setup() 
+{
+  Wire.begin();         // join i2c bus (address optional for master)
+  Serial.begin(115200); // start serial for output
+  while (Serial)
```
Comment (eriknyquist) [inline libraries/Wire/examples/master_check/master_check.ino]: Did you mean `while (!Serial);` ?

File: libraries/Wire/examples/master_check/master_check.ino [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+#define BUFFER_SIZE 8
+
+static int count = 0;  // recode the higher 4 bits of data
+static uint8_t buffer_sender[BUFFER_SIZE]; // data source for master writer
+static uint8_t buffer_receiver[BUFFER_SIZE]; // data distination for master reader
+
+void setup() 
+{
+  Wire.begin();         // join i2c bus (address optional for master)
+  Serial.begin(115200); // start serial for output
```
Comment (eriknyquist) [inline libraries/Wire/examples/master_check/master_check.ino]: Convention for example sketches is to use 9600 baud for Serial

File: libraries/Wire/examples/master_check/master_check.ino [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+void loop() 
+{
+  count++;
+  Wire.beginTransmission(8); // transmit to device #8
+
+  // Step 1:  Making the pattern.  Note that the count is in the upper 4 bit of the byte.
+  // Step 2:  And write it out to the device too.
+  for (int i = 0; i < BUFFER_SIZE; i++) 
+  {
+    buffer_sender[i] = ((count & 0xf) << 4) | i;
```
Comment (eriknyquist) [inline libraries/Wire/examples/master_check/master_check.ino]: Please keep the examples simple-- many Arduino users would be baffled by this line.
You can use the existing Wire example sketches as a reference, you will see they are very short and simple.

**Comment by eriknyquist** [comment]
My only issues are with the master_check.ino example, so if this will be removed like @bigdinotech says, then it's a +1 from me for everything else.

**Comment by SidLeung** [comment]
Example, master_check, is removed since it is more for testing not use application.  Bug fixed for closing the adapter.  Will squash the PR's.

**Comment by SidLeung** [comment]
Code been reviewed by @bigdinotech @eriknyquist @xieqi, corrections made accordingly.  Please merge @calvinatintel

**Comment by calvinatintel** [comment]
Cherry-picked into the 1.0.7 branch. Will be closed once 1.0.7 is final.

**Comment by calvinatintel** [comment]
cherry-picked

**Pull request closed without merging**